flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java

```diff
@@ -162,53 +162,57 @@ public void registerScalarFunction(String name, ScalarFunction function) {
 		FunctionDefinition userCandidate = null;
 
 		Catalog catalog = catalogManager.getCatalog(catalogManager.getCurrentCatalog()).get();
+		ObjectPath functionPath = new ObjectPath(catalogManager.getCurrentDatabase(), functionName);
 
-		if (catalog.getTableFactory().isPresent() &&
+		if (catalog.functionExists(functionPath)) {
+			if (catalog.getTableFactory().isPresent() &&
 				catalog.getTableFactory().get() instanceof FunctionDefinitionFactory) {
-			try {
-				CatalogFunction catalogFunction = catalog.getFunction(
-					new ObjectPath(catalogManager.getCurrentDatabase(), functionName));
+				try {
+					CatalogFunction catalogFunction = catalog.getFunction(functionPath);
 
-				FunctionDefinitionFactory factory = (FunctionDefinitionFactory) catalog.getTableFactory().get();
+					FunctionDefinitionFactory factory = (FunctionDefinitionFactory) catalog.getTableFactory().get();
 
-				userCandidate = factory.createFunctionDefinition(functionName, catalogFunction);
-			} catch (FunctionNotExistException e) {
-				// Ignore
-			}
-
-			return Optional.of(
-					new FunctionLookup.Result(
-						ObjectIdentifier.of(catalogManager.getCurrentCatalog(), catalogManager.getCurrentDatabase(), name),
-						userCandidate)
-				);
-		} else {
-			// Else, check in-memory functions
-			userCandidate = userFunctions.get(functionName);
+					userCandidate = factory.createFunctionDefinition(functionName, catalogFunction);
 
-			final Optional<FunctionDefinition> foundDefinition;
-			if (userCandidate != null) {
-				foundDefinition = Optional.of(userCandidate);
+					return Optional.of(
+						new FunctionLookup.Result(
+							ObjectIdentifier.of(catalogManager.getCurrentCatalog(), catalogManager.getCurrentDatabase(), name),
+							userCandidate)
+					);
+				} catch (FunctionNotExistException e) {
+					// Ignore
```

Comment:
Instead of ignoring the exception, I'm wondering if we should throw a TableException. This is because the function exists previously but now stops existing, which implies an unexpected scenario.